src/utils/mixins.js

```diff
@@ -192,53 +245,24 @@ const fileUploader = {
       this.errorMessage = null;
       const hookResult = await preUpload();
       this.indeterminate = false;
-      const promiseList = this.files.map((file) => {
-        let promiseChain = Promise.resolve();
-        if (file.status === 'done') {
-          // We are resuming, skip already completed files
-          return promiseChain.then(() => file.result);
-        }
-        const progress = (event) => {
-          this.$set(file, 'progress', event);
-        };
-        file.status = 'uploading';
-        file.progress.indeterminate = true;
-        if (file.upload) {
-          promiseChain = promiseChain.then(() => file.upload.resume());
-        } else {
-          // eslint-disable-next-line new-cap
-          file.upload = new uploadCls(file.file, {
-            $rest: this.girderRest,
-            parent: (hookResult && hookResult.dest) ? hookResult.dest : dest,
-            progress,
-            params: file.uploadClsParams,
-          });
-          promiseChain = promiseChain
-            .then(() => file.upload.beforeUpload())
-            .then(() => file.upload.start());
+      const promiseList = [];
+      // shallow copy into queue
+      const fileQueue = this.files.slice().reverse();
```

Comment:
Any reason to bother reversing it?

Reply:
`.pop()` dequeues from the end, so to process them in order, you have to reverse the list first.

Reply:
Ah, I had forgotten that the UI will be better if they go in order. Ideally we wouldn't even need to do a shallow copy or a reverse, and modifying the list each time isn't necessary -- we could just use a generator to iterate over the list and return each element in place. (If generators are too expensive to bother with, it could be an index variable in the closure...)

Reply:
Or, if you think all of this bikeshedding isn't worth it, let's at least remove the `reverse()` and use `shift()` instead of `pop()`.

Reply:
JS has no "atomic test and set", so using an index variable requires you to have a little more confidence in how the JS thread scheduler works. 

`const file = queue.pop()` gives me more confidence than

```
while (i < this.files.length) {
  const file = this.files[i++];
}
```
because I'm not 100% sure that those lines won't be interleaved.  (I am like 90% sure).

> we could just use a generator 

A generator would just get transpiled with regenerator-runtime, right?  I'm pretty sure we're already bundling it, but after it gets transpiled, the memory footprint will probably be the same.

Reply:
Practically, I like `shift()`, but this discussion is also interesting, so if you want to make an argument for one of the other approaches I want to hear it.

Reply:
> because I'm not 100% sure that those lines won't be interleaved

Those lines will definitely not be interleaved. JavaScript is run via a single thread using an event loop approach for cooperative multitasking. The unit of atomicity is the function itself. Once a function starts, nothing will interrupt it until it reaches the end of its execution and returns. (The exception to this rule is if the function calls `await`.) So there's no need to ever worry about thread safety of data in JS. Even if you take advantage of web workers, which allows the use of multiple threads, there are very strict data ownership semantics that make atomicity concerns moot.

From a pure readability and simplicity standpoint, I find the generator approach best. It provides the exact behavior we want with no copying. But what I'm not sure on is how nasty that is from a babel / polyfilling perspective. Modern browsers support them natively now, but I'm assuming Babel will transpile them away with our configuration.